### PR TITLE
4-letter search words

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

Link to YouTrack:
MANG-7775 4-letter search terms like "cubes" and "cones" still return nothing in search

#### Why does search give plural results in another order?
E.g. if we search for "cube". We have next result:
image

But if we search for "cubes". We have the same results, but in another order:
image

The answer is next:
BoundingBox.ToCuboid has next search tags: `converttosolid,bb2cube,bb2cub,bounding tosolid,cubes`.
Cuboid.ByCorners has next search tags: `box,cube,byminmax,by corners,by points,cubes`.

As you can see ByCorners fully match for the word "cube", while ToCuboid doesn't have such search term. But if we search for the word "cubes", then both of them have such term. And since BoundingBox goes before Cuboid (B before C), it appears first in search.

#### More flexible search
So, it's time to make our search more flexible. My proposal is to add for every search tag its' own weight. So that we could influence on node's place in search.

I've added new xml tag - `<weights></weights>`. Now we can add to ZeroTouch nodes search tags with weights.
E.g.
```xml
<member name="M:Autodesk.DesignScript.Geometry.BoundingBox.ToCuboid">
    <summary>
             Get the BoundingBox as a Solid Cuboid
    </summary>
    <search>
            converttosolid,bb2cube,bb2cub,bounding tosolid,cubes
    </search>
    <weights>
            0.5,0.4,0.3,0.5,0.3
    </weights>
</member>
```
If weight is not set, then by default it's set to 0.5 as it was before.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@Benglin 

### FYIs

@pboyer 

### For QA
Hi, @lillismith and @jnealb .
There is one thing, that still looks meaningless for me.

If we search for "arc". We get Arc nodes like: ByThreePoints, ByBestTroughPoints, ByFillet etc. After Geometry nodes there are Math nodes like: Acos, Asin, Atan etc. That's ok, because these nodes have next search tags: `arcsin`, `arccosine` and `arctangent`. But I don't see any reason why we should add them to plural search result!

The same thing for "face" and "faces". If we search for "face", we get: Face.Edges and Face.Vertices. If we search for "faces", we get: PolySurface.SurFACES. And in my opinion this is right search result. If we search of singular - we get Face. If we search for plural - we get PolySurface. This sounds more reasonable, then return the same results.

So, what I meant to say: please, look one more time though singular/plural results and think  which of them do we really need?

### CODE REVIEW

**Comment by jnealb** [comment]
@aosyatnik @pboyer @Benglin @lillismith Yes, agreed that slavish pluralization is incorrect. Question: can the intra 'Geometry - PolySurface-Actions' alphabetical sort be overridden to reflect the "faces" search, so that "polysurface.extractsolids" is not listed at the top, which appears effectively as the #2 hit (and is confusing until you ponder it: not good ui)? Or is that just a bad idea?

geomrank

File: src/DynamoCore/Search/SearchDictionary.cs [hunk 1]
```diff
@@ -101,8 +101,12 @@ public void Add(V value, IEnumerable<string> tags, double weight = 1)
...
                 OnEntryAdded(value);
             }
-            foreach (var tag in tags.Select(x => x.ToLower()))
-                keys[tag] = weight;
+
+            for (int i = 0; i < tags.Count(); i++)
+            {
+                var tag = tags.ElementAt(i).ToLower();
+                keys[tag] = weights.ElementAt(i);
+            }
```
Comment (Benglin) [inline src/DynamoCore/Search/SearchDictionary.cs]: `tags.Count()` will be called multiple times, which forces multiple enumeration on `tags` -- we can either cache the count value, or use a `foreach`. Also, this assumes that both `tags` and `weights` are of the same length, we need to handle unequal sized list cases.

**Comment by Benglin** [comment]
Hi @aosyatnik, the code changes look good to me. Thanks. The thing is I'd still need @pboyer to chime in here because we are addressing more than just weights of search terms (we're trying to solve the problem of using plural search terms). Peter may have something else in mind?

File: src/DynamoCore/Search/SearchDictionary.cs [hunk 2]
```diff
@@ -101,8 +101,13 @@ public void Add(V value, IEnumerable<string> tags, double weight = 1)
...
                 OnEntryAdded(value);
             }
-            foreach (var tag in tags.Select(x => x.ToLower()))
-                keys[tag] = weight;
+
+            int tagsCount = tags.Count();
+            for (int i = 0; i < tagsCount; i++)
+            {
+                var tag = tags.ElementAt(i).ToLower();
+                keys[tag] = weights.ElementAt(i);
```
Comment (aosyatnik) [inline src/DynamoCore/Search/SearchDictionary.cs]: Yes, @Benglin , you've noticed absolutely right. Length of `weights` should be the same as `tags`.

**Comment by Racel** [comment]
@jnealb - I think from a UX perspective that switching to rank order only (over having a top search result followed by an alphabetical list) makes so much more sense. Currently, it is very confusing that the top search results appears twice (once as the top result and again in the alphabetical list). But, I also think a change like this will require some time to define the proper UI fix/implementation. @jnealb @kronz @pboyer - what do you think about adding a future task to address this?

File: src/DynamoCore/Search/Interfaces/ISearchEntry.cs [hunk 3]
```diff
@@ -19,6 +19,11 @@ public interface ISearchEntry
         ICollection<string> SearchTags { get; }
 
         /// <summary>
+        ///     Every search tag should have weight. This weight will be taken into account during search.
+        /// </summary>
+        IEnumerable<double> TagsWeights { get; }
```
Comment (pboyer) [inline src/DynamoCore/Search/Interfaces/ISearchEntry.cs]: Should be `SearchTagWeights`

File: src/DynamoCore/Search/SearchElements/NodeSearchElement.cs [hunk 4]
```diff
@@ -18,6 +18,7 @@ public abstract class NodeSearchElement : ISearchEntry, ISource<NodeModel>
         protected string iconName;
 
         private readonly HashSet<string> keywords = new HashSet<string>();
+        protected readonly List<double> keywordsWeights = new List<double>();
```
Comment (pboyer) [inline src/DynamoCore/Search/SearchElements/NodeSearchElement.cs]: `keywordWeights` - In English, the adjective should not be plural, as that would be redundant. :)

File: extern/ProtoGeometry/ProtoGeometry.XML [hunk 5]
```diff
@@ -3550,7 +3584,8 @@
             <summary>
             Scale in one dimension by base and 2 pick points.  The scaling axis is defined by the line between base and pick0.
             </summary>
-            <search>resize,size,from,to,scale1d,1d</search>
+            <search>resize,size,from,to,scale1d,1d,lines</search>
+            <weights>0.5,0.5,0.5,0.5,0.5,0.5,0.4</weights>
         </member>
         <member name="M:Autodesk.DesignScript.Geometry.CoordinateSystem.Scale2D(Autodesk.DesignScript.Geometry.Plane,Autodesk.DesignScript.Geometry.Point,Autodesk.DesignScript.Geometry.Point)">
             <summary>
```
Comment (pboyer) [inline extern/ProtoGeometry/ProtoGeometry.XML]: Did you make this change in the closed source repository, too? We'll need to update it to transfer the weights.

**Comment by pboyer** [comment]
@Benglin @aosyatnik @lillismith,

These PR's worry me a lot because they feel like spot fixes. How do we verify that this test is "better" than the last? We need a way of quantifying this information.

I'm hesitant to merge this without hatching a plan for how to approach this. I'll return to this tomorrow.

**Comment by aosyatnik** [comment]
Yes, @pboyer , I agree with you about validation of results. Maybe, @lillismith  or @Racel  could create some sort of list, where there would be search words and assumed results. Then I could write test cases.

**Comment by Benglin** [comment]
The answer to that, @pboyer, is a system that learns, as you pointed out separately. For now I think the core issue of this is we are not able to handle *fuzzy* search, I'd expect *cube* and *cubes* to have the exact same result. More or less equivalent to `~` operator in Google search.

File: src/DynamoCore/Search/SearchDictionary.cs (hunk 2 above)
Comment (pboyer) [inline src/DynamoCore/Search/SearchDictionary.cs]: We should check that these two lists have the same length. If they don't, we should throw a descriptive exception.

File: src/DynamoCore/Search/Interfaces/ISearchEntry.cs [hunk 6]
```diff
@@ -19,6 +19,11 @@ public interface ISearchEntry
         ICollection<string> SearchTags { get; }
 
         /// <summary>
+        ///     Every search tag should have weight. This weight will be taken into account during search.
+        /// </summary>
+        IEnumerable<double> SearchTagWeights { get; }
```
Comment (pboyer) [inline src/DynamoCore/Search/Interfaces/ISearchEntry.cs]: `SearchTagWeights`

File: src/DynamoCore/Library/MemberDocumentNode.cs [hunk 7]
```diff
@@ -8,6 +8,7 @@ class MemberDocumentNode
         private readonly string fullyQualifiedName;
         private string summary = String.Empty;
         private string searchTags = String.Empty;
+        private string tagsWeights = String.Empty;
```
Comment (pboyer) [inline src/DynamoCore/Library/MemberDocumentNode.cs]: `searchTagWeights`

File: src/DynamoCore/Library/MemberDocumentNode.cs [hunk 8]
```diff
@@ -41,6 +42,20 @@ public string SearchTags
             }
         }
 
+        public string TagsWeights
```
Comment (pboyer) [inline src/DynamoCore/Library/MemberDocumentNode.cs]: `SearchTagWeights`

File: src/DynamoCore/Library/XmlDocumentationExtensions.cs [hunk 9]
```diff
@@ -42,6 +42,16 @@ public static IEnumerable<string> GetSearchTags(this FunctionDescriptor member,
                 .Where(x => x != String.Empty);
         }
 
+        /// <param name="xml">Don't set it, it's just for tests.</param>
+        public static IEnumerable<double> GetTagsWeights(this FunctionDescriptor member, XmlReader xml = null)
```
Comment (pboyer) [inline src/DynamoCore/Library/XmlDocumentationExtensions.cs]: `GetSearchTagWeights`

File: src/DynamoCore/Library/XmlDocumentationExtensions.cs [hunk 10]
```diff
@@ -234,6 +247,7 @@ private enum XmlTagType
             Summary,
             Parameter,
             SearchTags,
+            TagsWeights,
```
Comment (pboyer) [inline src/DynamoCore/Library/XmlDocumentationExtensions.cs]: `SearchTagWeights`

File: src/DynamoCore/Library/XmlDocumentationExtensions.cs [hunk 11]
```diff
@@ -241,7 +255,8 @@ private enum DocumentElementType
         {
             Summary,
             Description,
-            SearchTags
+            SearchTags,
+            TagsWeights
```
Comment (pboyer) [inline src/DynamoCore/Library/XmlDocumentationExtensions.cs]: `SearchTagWeights`

**Comment by Racel** [comment]
@aosyatnik @jnealb @Benglin  - In the latest build, the plural and singular search results do not match for several terms (arc(s), face(s), map(s), rank(s) | In Revit: mesh(es), UV(s)). @jnealb requested that these search results for plural and singular terms match, and I agree. I think all of the nodes that appear in the singular results should appear in the plural results and vice versa. The only difference should be the order of the results. 

The results should probably be ordered based on the frequency of use (but we won't be able to determine this  until we get the results back from the search logs that Luke is collecting).

**Pull request merged**